src/cudnn/tensor/operation/reduce.rs

```diff
@@ -0,0 +1,379 @@
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::ptr::null_mut;
+
+use alloc::rc::Rc;
+
+use crate::cudnn::sys::*;
+use crate::driver::sys::{cuMemAllocAsync, cuMemFreeAsync, CUdeviceptr};
+use crate::prelude::*;
+
+pub struct ReduceDescriptor(cudnnReduceTensorDescriptor_t);
+
+impl ReduceDescriptor {
+    pub fn create() -> CudaCudnnResult<Self> {
+        let mut descriptor = MaybeUninit::uninit();
+        unsafe {
+            cudnnCreateReduceTensorDescriptor(descriptor.as_mut_ptr()).result()?;
+            Ok(Self(descriptor.assume_init()))
+        }
+    }
+}
+impl Drop for ReduceDescriptor {
+    fn drop(&mut self) {
+        unsafe { cudnnDestroyReduceTensorDescriptor(self.0) }
+            .result()
+            .unwrap();
+    }
+}
+pub struct TensorReduction<
+    O,
+    T,
+    const N_IN: usize,
+    const C_IN: usize,
+    const H_IN: usize,
+    const W_IN: usize,
+    const N_OUT: usize,
+    const C_OUT: usize,
+    const H_OUT: usize,
+    const W_OUT: usize,
+> {
+    descriptor: ReduceDescriptor,
+    op: PhantomData<O>,
+    data_type: PhantomData<T>,
+    device: Rc<CudaDevice>,
+    a: Rc<TensorDescriptor<T, N_IN, C_IN, H_IN, W_IN>>,
+    out: Rc<TensorDescriptor<T, N_OUT, C_OUT, H_OUT, W_OUT>>,
+    workspace_allocation: CUdeviceptr,
+    workspace_size: usize,
+}
+pub trait ReduceOp {
+    fn get_op() -> cudnnReduceTensorOp_t;
+}
+macro_rules! impl_reduce_op {
+    ($struct:ident : $op:ident) => {
+        pub struct $struct;
+        impl ReduceOp for $struct {
+            fn get_op() -> cudnnReduceTensorOp_t {
+                cudnnReduceTensorOp_t::$op
+            }
+        }
+    };
+}
+impl_reduce_op!(ReduceOperationAdd: CUDNN_REDUCE_TENSOR_ADD);
+impl_reduce_op!(ReduceOperationMul: CUDNN_REDUCE_TENSOR_MUL);
+impl_reduce_op!(ReduceOperationMin: CUDNN_REDUCE_TENSOR_MIN);
+impl_reduce_op!(ReduceOperationMax: CUDNN_REDUCE_TENSOR_MAX);
+impl_reduce_op!(ReduceOperationAMax: CUDNN_REDUCE_TENSOR_AMAX);
+impl_reduce_op!(ReduceOperationAvg: CUDNN_REDUCE_TENSOR_AVG);
+impl_reduce_op!(ReduceOperationNorm1: CUDNN_REDUCE_TENSOR_NORM1);
+impl_reduce_op!(ReduceOperationNorm2: CUDNN_REDUCE_TENSOR_NORM2);
+impl_reduce_op!(ReduceOperationMulNoZeros: CUDNN_REDUCE_TENSOR_MUL_NO_ZEROS);
```

Comment:
Yeah this will all be super useful. Let's move to using runtime shapes though as mentioned in other places, I can help with this